concept/type_/mod.rs

```diff
@@ -136,34 +129,117 @@ pub trait TypeAPI<'a>: ConceptAPI<'a> + TypeVertexEncoding<'a> + Sized + Clone +
         type_manager: &'m TypeManager,
     ) -> Result<MaybeOwns<'m, Vec<Self>>, ConceptReadError>;
 
+    fn get_supertype_root<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+    ) -> Result<Option<Self>, ConceptReadError> {
+        Ok(self.get_supertypes_transitive(snapshot, type_manager)?.into_iter().cloned().last())
+    }
+
     fn get_subtypes<'m>(
         &self,
         snapshot: &impl ReadableSnapshot,
         type_manager: &'m TypeManager,
-    ) -> Result<MaybeOwns<'m, Vec<Self>>, ConceptReadError>;
+    ) -> Result<MaybeOwns<'m, HashSet<Self>>, ConceptReadError>;
 
     fn get_subtypes_transitive<'m>(
         &self,
         snapshot: &impl ReadableSnapshot,
         type_manager: &'m TypeManager,
     ) -> Result<MaybeOwns<'m, Vec<Self>>, ConceptReadError>;
+
+    fn is_supertype_of<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+        other: Self,
+    ) -> Result<bool, ConceptReadError> {
+        Ok(other.get_supertype(snapshot, type_manager)?.eq(&Some(self.clone())))
+    }
+
+    fn is_supertype_transitive_of<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+        other: Self,
+    ) -> Result<bool, ConceptReadError> {
+        Ok(other.get_supertypes_transitive(snapshot, type_manager)?.contains(self))
+    }
+
+    fn is_supertype_transitive_of_or_same<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+        other: Self,
+    ) -> Result<bool, ConceptReadError> {
+        Ok(self == &other || self.is_supertype_transitive_of(snapshot, type_manager, other)?)
+    }
+
+    fn is_subtype_of<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+        other: Self,
+    ) -> Result<bool, ConceptReadError> {
+        Ok(other.get_subtypes(snapshot, type_manager)?.contains(self))
+    }
+
+    fn is_subtype_transitive_of<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+        other: Self,
+    ) -> Result<bool, ConceptReadError> {
+        Ok(other.get_subtypes_transitive(snapshot, type_manager)?.contains(self))
+    }
+
+    fn is_subtype_transitive_of_or_same<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+        other: Self,
+    ) -> Result<bool, ConceptReadError> {
+        Ok(self == &other || self.is_subtype_transitive_of(snapshot, type_manager, other)?)
+    }
+
+    fn inheritance_cmp<'m>(
+        &self,
+        snapshot: &impl ReadableSnapshot,
+        type_manager: &'m TypeManager,
+        other: Self,
```

Comment:
What's the expected 'inheritance_cmp' if they're not equal but in siblings? :D

Reply:
Don't quite get the case of 

> not equal but in siblings

"siblings" (based on my unofficial and potentially banned notation) are interface types with the same root parent and with the same object type. And I actually do not care about "siblings" in this method. 

You might be confused by `std::cmp::Ordering::Equal` in case if types are not equal, and I could return "None" in this case. However, comparators are usually simple like that, and used for ordering, not for hashing to store in buckets. Sooo we'd just leave these types somewhere together, surrounded by `Less` and `Greater`, which is the point of sorting. Might have bad side effects if misused, maybe there is a better way to express it.
It is used in `type_get_constraints_closest_source` in `constraint.rs`

Reply:
Address

Reply:
Done